craft_parts/state_manager/state_manager.py

```diff
@@ -0,0 +1,152 @@
+# -*- Mode:Python; indent-tabs-mode:nil; tab-width:4 -*-
+#
+# Copyright (C) 2021 Canonical Ltd.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Part crafter step state management."""
+
+import itertools
+from dataclasses import dataclass
+from typing import Dict, Optional
+
+from craft_parts.steps import Step
+
+from .states import StepState
+
+
+@dataclass(frozen=True)
+class _StateWrapper:
+    """A wrapper for the in-memory StepState class with extra metadata.
+
+    This is a wrapper class for StepState that adds additional metadata
+    such as the update sequence order to the data loaded from a previous
+    lifecycle run. Metadata is read-only to prevent unintentional changes.
+    """
+
+    state: StepState
+    serial: int
+    step_updated: bool = False
+
+    def is_newer_than(self, other: "_StateWrapper"):
+        """Verify if this state is newer than the specified state.
+
+        :param other: The wrapped state to compare this state to.
+        """
+        return self.serial > other.serial
+
+
+class _EphemeralStates:
+    """A dictionary-backed simple database manager for wrapped states."""
+
+    def __init__(self):
+        self._state: Dict[str, Dict[Step, _StateWrapper]] = {}
+        self._serial_gen = itertools.count(1)
+
+    def wrap_state(self, state: StepState, step_updated: bool = False) -> _StateWrapper:
+        """Add metadata to step state.
+
+        :param state: The part state to store.
+        :param step_updated: Whether this state was updated after an
+            outdated report.
+
+        :return: The wrapped state with additional metadata.
+        """
+        stw = _StateWrapper(
+            state, serial=next(self._serial_gen), step_updated=step_updated
+        )
+        return stw
+
+    def set(
```

Comment:
Perhaps be a little clearer:

```suggestion
    def set_part_step_state()
```

Reply:
These methods are only used by the `StateManager` class, and the usage should be clear there (something like `self._state_db.set(part_name=part.name, step=step, state=state)`). If not, we can refactor after examining the context in `StateManager`.